xarray/core/groupby.py

```diff
@@ -431,12 +434,32 @@ def __len__(self):
     def __iter__(self):
         return zip(self._unique_coord.values, self._iter_grouped())
 
+    def weighted(self, weights):
+        """Set weights for GroupBy reductions."""
+        self._weighted = True
```

Comment:
Let's try to make a new `WeightedGroupBy` object rather than mutating the existing GroupBy object in-place.